rfc/002-fault-tolerance/README.md

```diff
@@ -0,0 +1,426 @@
+that requires an ACK, which needs to travel all the way back to the source of
+the input stream so that the event stream can be advanced.
+
+TODO: describe how we can do this with CAF as of today.
+
+In technical terms, we can extend the CAF Streaming message type to include a
+handle to the source that requires ACK'ing, and then trigger an ACK once the
+data has been persisted. VAST already implements a similar mechanism today that
+acknowledges when events arrived at an active partition, which (almost)
+guarantees that they are queryable.
+
+### Read Path
+
+For historical queries, we can achieve fault-tolerance on the read part by
+moving query execution into a dedicated process. This requires that the catalog
+is the sole owner of partitions, and that the catalog returns the set of
+candidate partitions relevant to answer a query.
+
+Live queries bypass the catalog and would therefore need to implement ACK'ing
+independently, at the granularity of table slices. We do not cover this process
+in detail, as it constitutes a subset of the ACK'ing for the write path outlined
+above.
+
+### Metastore
+
+The metastore primarily stores the state of partition data. We restrict our analysis to this use case. From the perspective of the catalog, the metastore API should expose itself as an actor that encapsulates the implementation details. This actor can rely on VAST's plugin architecture to implement the metastore-specific integration.
+
+#### Request-Response API
+
+The catalog requires the following interface from the metastore:
+
+1. `get: [uuid] → [partition]`
+2. `add: [partition] → ACK`
+3. `delete: [uuid] → ACK`
+4. `replace: [uuid], [partition] → ACK`
+
+The semantics of the mentioned types are as follows:
+
+- `partition`: metadata about a partition, such as its URL and its schema.
+- `uuid`: a globally unique identifier for a partition.
+- `ACK`: a confirmation about an action.
+
+The semantics the function signatures are as follows:
+
+1. `get`: retrieves a list of `partition` objects corresponding one-to-one to
+   the requested list of partition UUIDs.
+
+2. `add`: registers a list of new partitions.
+
+3. `delete`: removes a list of partition from the metastore (doesn't delete the
+   underlying data)
+
+4. `replace`: performs an atomic deletion and addition of partition data, by
+   deleting the given UUIDs and adding new partition metadata.
+
+In theory, the functionality of `replace` is a superset of `add` and `delete`.
+For ease of exposition, we kept them as separate API functions.
+
+#### Subscription API
+
+The metastore must also expose a subscription mechanism to hook into the feed of
+updates, such that the catalog can learn about new partitions from other catalog
+instances.
+
+Once a catalog learns about a new partition, it will load its sketches so that
+they become available for querying.
+
+Given these requirements, we briefly looked at the following systems with regard
+to their suitability.
+
+#### Candidates
```

Comment:
I understand that you're still researching, once you have a clear idea, it would be good to clearly say which candidate you recommend. I don't feel I have enough information to fully review the proposal until the trade-offs are clearer.

Reply:
I don't think we will fully commit to one specific solution in the scope of this RFC, although I think we can narrow it down to "native RAFT" or etcd. I would still keep the preliminary evaluation of the other alternatives to show that those were also considered.
As etcd is already battle tested I think it makes sense to give it the recommendation.

Reply:
I incorporated that into the proposal with [db37533](https://github.com/tenzir/vast/pull/2562/commits/db37533243683055f01b4eb34da35c99a10da563).

Reply:
Thanks, that makes sense and helps me :-) (feel free to close this thread)